enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java

```diff
@@ -337,37 +336,174 @@ public void registerFileListing( NeoStoreFileListing fileListing )
         fileListing.registerStoreFileProvider( this::snapshotStoreFiles );
     }
 
-    private Resource snapshotStoreFiles( Collection<StoreFileMetadata> files ) throws IOException
+    @Override
+    public void awaitFlip()
     {
-        final Collection<ResourceIterator<File>> snapshots = new ArrayList<>();
-        for ( WritableFulltext index : Iterables.concat( writableNodeIndices.values(), writableRelationshipIndices.values() ) )
+        int startSide = tentativeSide;
+        synchronized ( this )
+        {
+            while ( startSide == tentativeSide || side != tentativeSide )
+            {
+                try
+                {
+                    wait();
+                }
+                catch ( InterruptedException e )
+                {
+                    Thread.currentThread().interrupt();
+                    return;
+                }
+            }
+        }
+    }
+
+    private void flip() throws IOException
+    {
+        configurationLock.writeLock().lock();
+        try
         {
-            // Save the last committed transaction, then drain the update queue to make sure that we have applied _at least_ the commits the config claims.
-            index.saveConfiguration( transactionIdStore.getLastCommittedTransactionId() );
+            List<AsyncFulltextIndexOperation> populations = new ArrayList<>();
+            synchronized ( this )
+            {
+                int currentSide = side;
+                tentativeSide = (currentSide + 1) % 2;
```

Comment:
No need to have the `currentSide` here, also, this looks strange since we have `A_SIDE` and `B_SIDE`. How about
```
tentativeSide = side == A_SIDE ? B_SIDE : A_SIDE;
```